### DISCUSSION TITLE
Raspberry Pico and MCP23017 randomly not working

### DISCUSSION THREAD

**Discussion by dmilardovich** [micropython/micropython#12480] [RP2040 / Pico]
Hi all, 

 I am using a Raspberry Pi Pico, programmed in MicroPython with Thonny and connected via I2C to an MCP23017 IO expander. It works OK, but every 3-5 minutes, it randomly stops working, with the following error: 

 Traceback (most recent call last):
  File "<stdin>", line 42, in <module>
  File "mcp23017_new.py", line 80, in value
  File "mcp23017_new.py", line 43, in bit
  File "mcp23017_new.py", line 26, in read
OSError: [Errno 5] EIO

 I made a new circuit in a breadboard, with just the Raspberry Pi Pico and the MCP23017, and a very simple script, as shown next: 

 from machine import I2C, Pin, RTC, ADC, SPI, UART
import mcp23017_new

# Set-up the I2C bus.
i2c_bus = I2C(1,
              scl = Pin(19, pull=Pin.PULL_UP),
              sda = Pin(18, pull=Pin.PULL_UP),
              freq = 10_000)

# Set-up the MCP23017 I/O expander.
IO_expander = mcp23017_new.MCP23017(i2c_bus, 0x20)

while (True):
    button_UP = IO_expander.pin(0).value()
    print(" ")
    print(str(button_UP))

 And the problem persists. I already tried lowering the I2C frequency, using a different driver, setting the I2C pins with pull-up resistors and placing physical 3.3 kOhm pull-up resistors in the breadboard. None of this solved the issue. 

 Could you pleas tell me where the problem might be? 

 Thanks! 
 Best, 
 Diego

**Comment by Lobo-T** [comment]
Not really enough information to help you so this is just a shot in the dark.  Have you connected all the address pins to either ground or vcc?  Thet are not internally pulled.  I forgot them on one pcb design, four of the five cards still worked.

**Comment by mcauser** [comment]
Hi, I assume you are referring to this: https://github.com/mcauser/micropython-mcp23017/issues/16

OSError 5 means your Pico can't find / communicate with the MCP.
Is it the MCP reliably showing up under `i2c.scan()`?
Could try larger pull-ups (4.7-10k), changing the I2C bus frequency, trying SoftI2C.
Standard 100k, 400k speeds are mentioned in the MCP datasheet.
Maybe add a small `sleep()` in your while loop.

All of my testing was on an ESP32 board, using I2C defaults.
I've noticed many projects report OSError 5's on Picos.
I'm not sure what the root cause is or the fix...

Yes, the 3x address select pins need to be pulled to either VCC/GND, which gives you an I2C address in the range 0x20-0x27 (32-39).

**Comment by dmilardovich** [comment]
Hi, thanks both for your answers. 

 I fixed the issue buy using a very simple function that checks if the connection between the MCP23017 and the Raspberry works OK and just re-tries to connect if it doesn't: 

     def IO_expander_update(self):
        self.IO_expander_updated = False

        while (self.IO_expander_updated == False):
            try:        
                # Read the buttons state. 
                self.button_UP = self.IO_expander.pin(0) 
                self.button_DOWN = self.IO_expander.pin(1)
                self.button_MODE = self.IO_expander.pin(2)
                self.button_PULSE = self.IO_expander.pin(3)
                self.button_DIS_ARM = self.IO_expander.pin(4)
                
                self.IO_expander_updated = True

            except OSError as e:
                    print("Error communicating with MCP23018. Re-trying.")
                    self.IO_expander_updated = True
                    pass

 However, I would still like to understand why the MCP23017 is failing at random intervals of time. Please find next answers to your questions and more information on the setup: 

 - All address pins are directly connected to GND. 
 - The reset pin is directly connected to VDD.
 - I tried using different pull-up resistors (the pull-up resistors of the Raspberry, which I think are about 30 kOhm, and also physical resistances with lower values: 3.3 KOhm and 480 Ohm). 
 - The i2c.scan() works OK and it shows the right address, considering that the 3 address pins are set to GND (0x20). 
 - I tried with higher and lower frequencies: 400 kHz and 10 kHz. 
 - I did not use sleep() in the test code I showed here, but I did use it in the "real" code. 
 - I tried using a capacitor between VDD and GND, right next to the 23017 pins, first 100 nF and then 200 nF. 
 - I tried with 3 different MCP23017 ICs. Same results for all of them. 
 
 None of that solved the problem. It seems the MCP23017 stops working for very short periods of time (I would guess in the us range) every 1-10 minutes. Any ideas on what might be causing this issue? 

 Thanks again!

**Reply by Lobo-T** [reply]
Is your program doing anything else at that interval?
Something that takes more power than usual?  Something that can block?

**Reply by dmilardovich** [reply]
No, the testing program is the one I posted in the first message. It does nothing else than checking the MCP23017.

**Comment by peterhinch** [comment]
The code above looks wrong. If you want it to keep trying the last two lines should be deleted:
```python
 def IO_expander_update(self):
    self.IO_expander_updated = False

    while (self.IO_expander_updated == False):
        try:        
            # Read the buttons state. 
            self.button_UP = self.IO_expander.pin(0) 
            self.button_DOWN = self.IO_expander.pin(1)
            self.button_MODE = self.IO_expander.pin(2)
            self.button_PULSE = self.IO_expander.pin(3)
            self.button_DIS_ARM = self.IO_expander.pin(4)
            self.IO_expander_updated = True
        except OSError as e:
            print("Error communicating with MCP23018. Re-trying.")
```
I can't think of a reason for the failure. It might be instructive to try a different platform e.g. ESP32.

**Reply by dmilardovich** [reply]
Thanks for your answer. The code works OK as it is. The last two lines are there because I do not want to keep trying, I just want to read the IO expander inputs once when I call the function.

**Reply by peterhinch** [reply]
OK, but in this case what is the purpose of the while loop? It will never run more than one iteration.

**Reply by dmilardovich** [reply]
The purpose is to read the IO expander pins just once. This is a function which is part of a larger code, it should read the IO expander pins just once each time it is used, not run indefinitely.

**Comment by bixb922** [comment]
Does using SoftI2C instead of I2C make any difference?

**Reply by dmilardovich** [reply]
Thanks. I never used SoftI2C. Is there any reason why that might work when the I2C fails?

**Reply by bixb922** [reply]
I asked out of curiosity and own interest. I am using Mcp23017 with esp32 and softi2c. With softi2c, timing is different, so there might be a different behaviour. I believe you also can use other frequencies with Softi2c.
(edit) With SoftI2C you can try very low freq= values, lower THAN hardware I2C (at least that is do on the ESP32). May be worth a try. If low frequencies fail: check if the cabling has no flaky connection. How long is the I2C cable? Could there be interference or EM noise?

**Reply by dmilardovich** [reply]
Thanks, I will have a look at softi2c. I already tried with a lower frequency of 10 KHz and that did not fix the issue. The cables are standard jumper-wires, I would say roughly 10 cm long. There should be no EM noise, since the breadboard is on my desktop in a regular office environment.

**Reply by bixb922** [reply]
If that's the context I'd take a look at the signals with a oscilloscope
